libcontainer/init_linux.go

```diff
@@ -134,6 +135,32 @@ func populateProcessEnvironment(env []string) error {
 	return nil
 }
 
+// verifyCwd ensures that the current directory is actually inside the mount
+// namespace root of the current process.
+func verifyCwd() error {
+	// getcwd(2) on Linux detects if cwd is outside of the rootfs of the
+	// current mount namespace root, and in that case prefixes "(unreachable)"
+	// to the returned string. glibc's getcwd(3) and Go's Getwd() both detect
+	// when this happens and return ENOENT rather than returning a non-absolute
+	// path. In both cases we can therefore easily detect if we have an invalid
+	// cwd by checking the return value of getcwd(3). See getcwd(3) for more
+	// details, and CVE-2024-21626 for the security issue that motivated this
+	// check.
+	//
+	// We have to use unix.Getwd() here because os.Getwd() has a workaround for
+	// $PWD which involves doing stat(.), which can fail if the current
+	// directory is inaccessible to the container process.
+	if wd, err := unix.Getwd(); errors.Is(err, unix.ENOENT) {
+		return errors.New("current working directory is outside of container mount namespace root -- possible container breakout detected")
+	} else if err != nil {
```

Comment:
Isn't there a legit case for which `unix.Getwd()` would return `ENOENT` that doesn't necessarily involve a breakout attempt? For example: `docker exec -it -w /non-existent-path`?

IOW, don't we need a separate if-statement to return a different error?

Reply:
No; if the working dir path does not exist, sysbox-runc would create it a bit earlier (see [here](https://github.com/nestybox/sysbox-runc/blob/d801147e695450fa8f651053b1c94ef2189762fb/libcontainer/rootfs_linux.go#L132)). Thus, by the time we get to this check, the path must exist and be accessible within the container.